src/parser/mxl/lomse_mxl_analyser.cpp

```diff
@@ -7765,6 +7776,23 @@ void MxlAnalyser::clear_pending_relations()
 
     m_lyrics.clear();
     m_lyricIndex.clear();
+    m_pendingDynamicsMarks.clear();
+}
+
+//---------------------------------------------------------------------------------------
+void MxlAnalyser::attach_pending_dynamics_marks(ImoNoteRest* pNR)
+{
+    for (ImoDynamicsMark* pDynamics : m_pendingDynamicsMarks)
+    {
+        ImoObj* pOldParent = pDynamics->get_parent_imo();
```

Comment:
This line needs to be changed to
```
ImoObj* pOldParent = pDynamics->get_parent_imo();       //this is the ImoAttachments child in the parent ImoContentObj (the ImoDirection obj)
pOldParent = pOldParent->get_parent_imo();     //this is the ImoContentObj (ImoDirection)   
```

Reply:
As far as I understand, dynamics marks are currently children of `ImoAttachments` in the internal model tree so this code should currently do a correct thing detaching dynamics mark from its parent `ImoAttachments`.

I agree however that it would be better to remove dynamics from `ImoDirection` by something like `remove_attachment()` call, thus making it symmetrical to the way it has been added there. However I see two issues here:
1. `remove_attachment()` function [deletes](https://github.com/lenmus/lomse/blob/d4537fb5e2c5397ffdfd50504364897d00be0094/src/internal_model/lomse_internal_model.cpp#L2153) the object it detaches. This is not what we want here so something different would be needed for this purpose (I see few `remove_but_not_delete...` functions across the code, possibly something like this would work here).
2. Finding `ImoDirection` by two `get_parent_imo()` is less obvious to understand when reading the code, and relevant changes in the internal model structure may break this code in a way which will be easy to miss. Therefore I think it would be preferable to use something like [`get_contentobj_parent()`](https://github.com/lenmus/lomse/blob/d4537fb5e2c5397ffdfd50504364897d00be0094/src/internal_model/lomse_internal_model.cpp#L662) in such context.

For now I have pushed [these changes](https://github.com/lenmus/lomse/compare/86150bb4427063e006aeb0ac1bbe799dcfddf5a9..1009add1961253d554f17a617081612e9ea6b98e) which should address the listed considerations.

Reply:
Thank you very much for your effort to improve the code. Both changes, `get_contentobj_parent()` and `remove_but_not_delete_attachment()` are certainly very well thought. The internal model needs some refactoring. Perhaps the next task I should do is to start with it.